### PR TITLE
Fix InMemoryFilmRepository concurrency issues

### PR DESCRIPTION
**Pull request by marcinmajkowski**
For `InMemoryFilmRepository` class to become thread-safe, `map` field is required to be `final` so safe-publication is ensured. Additionaly there was a race condition in `findAll` method, since `map` could change between calls to `values()` and `size()`.

### CODE REVIEW

**Comment by jakubnabrdalik** [comment]
Hi @marcinmajkowski 
Can you elaborate on why final is ensuring thread-safety? My understanding (possibly wrong) so far is that
- map is encapsulated inside InMemoryFilmRepository, thus the reference cannot be changed (no modifiers) without violating the encapsulation
- final does not guarantee that the reference won't change unfortunately (as pointed out by Douglas Hawkins in Concurrency Concepts in Java at Devoxx.pl 2017: https://www.youtube.com/watch?v=ADxUsCkWdbE)

**Comment by marcinmajkowski** [comment]
The necessity to use `final` in this case relates to the concept of _safe publication_. You can read about final Field Semantics here: http://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5 . This part is particularly important:

> An object is considered to be completely initialized when its constructor finishes. A thread that can only see a reference to an object after that object has been completely initialized is guaranteed to see the correctly initialized values for that object's final fields.

What is more important, is that it is not guaranteed that other threads will see correctly initialized field if it is not `final` (nor `volatile` or access to it is synchronized with other means). So in this particular case, it is possible (assuming InMemoryFilmRepository is accessed from multiple threads) that `map` can be seen as `null`.

This particular property of `final` is also mentioned in the presentation you linked (this presentation is great by the way) at around minutes 8 and 27.

**Comment by jakubnabrdalik** [comment]
Thanks @marcinmajkowski for explanation. 
Can you verify my line of thought once more though? 

From the spec you mentioned, and the presentation, the only way to the lack of final can hurt us, is if the reference to the object being constructed is seen by another thread before the object's constructor is finished.
This object has an empty constructor. Even if it had a custom one, I can safely assume it would not have any additional threads started/used inside the constructor. If such an action would be needed, it would be done in a Creator object from [GRASP](https://en.wikipedia.org/wiki/GRASP_(object-oriented_design)) which in case of Spring is usually the @Configuration class or a dedicated custom factory. Thus, this final is not needed.

If my line of thought is correct, please remove the final from the property (changes in findAll are very welcome). If it doesn't add any value, it shouldn't clutter the code.

**Comment by marcinmajkowski** [comment]
**Part V: Finals | Test Your Basic Understanding** from https://shipilev.net/blog/2014/jmm-pragmatics/#_happens_before_publication might be helpful. Basically you have no guarantee that InMemoryFilmRepository will be _safe-published_ (which would ensure other threads will see `map` only after its initialized) so you should _safe-initialize_ it by assigning to `final` field in constructor. This assignment will cause emission of trailing barrier at the end of constructor (can be verified here: http://hg.openjdk.java.net/jdk9/jdk9/hotspot/file/ae45df3285c9/src/share/vm/opto/parse1.cpp#l930 ) thus ensuring that assignment to `map` is visible for other threads. You can read about the topic here: https://shipilev.net/blog/2014/safe-public-construction/ .

Concurrency aside, using `final` by default is just a good practice I learn reading e.g. _Effective Java_, _Java Concurrency in Practice_ and watching numerous presentations on the topic. And it's not that I blindly follow what is said there. Their reasoning seems very reasonable to me.

I'm glad you challenge my proposal like that since it makes me research the topic deeper.

**Comment by jakubnabrdalik** [comment]
Hmmm... that's strange @marcinmajkowski. You wrote an answer but not to my question. Perhaps I worded it wrong. Let me try again:

**Since I do not need a guarantee that InMemoryFilmRepository will be safe-published, because I control the the controller, and will never use any other threads inside of it (that would be a design smell), is there any benefit in adding "final" to the field?**

I disagree that "using final by default" is a good practice. The reasons being:
- final does not protect the against changing the reference as pointed out by Douglas Hawkins above
- final as a "default" clutters code
- if this "default" was always true, I'd rather write an Aspect to add it in compile time, and not clutter the code

Looking at Effective Java (2002 edition) it does talk about final in terms of public fields (for example, page 49). And then private final is discussed:
"A final field containing a reference to a mutable object has all the disadvantages of a nonfinal field." (page 50). 
This map is mutable by design. That's the whole point of this class, thus adding a final there makes no sense.

There is a suggestion to "Make all fields final" (page 51) but that is in the context of immutability. Reasons being  concurrency (as discussed above), and revealing intention.

I'd argue, that in case of Spring Beans (by default singletons) this revealing is not needed. It's the same as checking for nullable values in Spring Beans constructors - while the practice itself makes sense (offensive programming), if you ever have a developer passing an explicit null to a constructor of a Spring Bean, you have a much bigger (though basic) problem in your project.

**Comment by marcinmajkowski** [comment]
Maybe I focused so much on thread safety of this class that I overlooked the context it is used in. It is a pretty common to not use `final` nor `volatile` for such Spring bean properties. But it is questionable whether it is a good practice. There are a couple of great answers here https://stackoverflow.com/questions/23906808/should-i-mark-object-attributes-as-volatile-if-i-init-them-in-postconstruct-in and the conclusion is, it would be highly appreciated to use `final` in such case. Due to Spring internals there is currently no threat in leaving `final` out in this particular case, but this fact is not documented ( https://jira.spring.io/browse/SPR-4307 ). What is documented is the fact that your bean is expected to be thread safe, and without this `final` it is not.

There is also this article on the subject: https://web.archive.org/web/20080226065034/http://blog.xebia.com/2007/03/01/spring-and-visibility-problems/ . Despite of how old this article is, it seems to me that the general concept still applies.

And finally, "Every single unnecessary character in an example app will be adopted as the de facto way to make stuff work in real projects", by Craig Walls. Although in this case it might be rather an unnecessary lack of certain characters :).

**Comment by jakubnabrdalik** [comment]
Good that we are on the same page then :)

Since multithreading in a constructor of a repository (or any other infrastructure service) is a code smell, I don't need to protect it. Just like JVM stops optimising/protecting us, if what we do is clearly wrong.

Since I don't want people to add final by default to every possible property assigned in constructor without understanding why (and I've seen it happen), I cannot put it here. 

Thanks for those link. Will come handy when I have to explain this to someone.

**Comment by jakubnabrdalik** [comment]
Not sure if it's interesting to you, but there is trick we've found in Lombok (although experimental), to have both safety and no additional code.

```
@FieldDefaults(makeFinal = true, level = PRIVATE)
class InMemoryFilmRepository implements FilmRepository {
    ConcurrentHashMap<String, Film> map = new ConcurrentHashMap<>();
```

**Comment by marcinmajkowski** [comment]
I'm not sure what you mean by multithreading in a constructor. The issue `final` lets you avoid in this case relates to client code being multithreaded, which is the case here since framework code instantiates and shares this repository class. Without `final` it is possible that after one thread instantiates `InMemoryFilmRepository` and somehow shares it with other thread, `map` value can be seen as `null`. It has nothing to do with whether `InMemoryFilmRepository` leverage multithreading in its constructor or not.

`final` is best used to express design intentions. Here the design of this repository is that `map` doesn't change after its initialized. When its present, I don't need to wonder whether `map` is reassigned while reading your code. Still it's not much of value in case of such simple class, but I like to simplify my approach to coding by applying best practices all the time. I don't like to think too much about whether my class is big/complex enough to start making it easier to reason about its internals.

**Comment by jakubnabrdalik** [comment]
Marcin I think we are running in circles now.

> Without final it is possible that after one thread instantiates InMemoryFilmRepository and somehow shares it with other thread

This "somehow" is what I'd call a design error. Have I ever seen anybody do it? No. Would I warn them on a review not to do it? Yes. Does it depend on the framework used? No.

> final is best used to express design intentions

Let me answer with a quote from Effective Java

> A final field containing a reference to a mutable object has all the disadvantages of a nonfinal field. 

This is a mutable object. Final gives you nothing, except false suggestions.

**Comment by marcinmajkowski** [comment]
But what I described is how Spring works. It instantiates your class and then multiple threads (e.g. separate thread per request) access it concurrently. It does not matter whether you leverage multithreading in your application explicitly.

According to the quote, I gave you a real example of an advantage of using `final` in this case. Yes, referenced object is still mutable, but with `final` you know that its always the same object. At least until you introduce errors related to safe publication in code executed concurrently.

**Comment by jakubnabrdalik** [comment]
> [Spring] instantiates your class and then multiple threads (e.g. separate thread per request) access it concurrently. It does not matter whether you leverage multithreading in your application explicitly.

All those other threads touch this object only after the constructor has finished. The problem described above is present only if other threads have access to the object before constructor finishes. That is, if the constructor uses/creates threads. Which is not what you should do anyway. Beans in Spring are accessible only after the constructor finishes. 
If I'm wrong, please point me to the code in Spring that behaves differently.

> Yes, referenced object is still mutable, but with final you know that its always the same object.

By the presentation above, it is not guaranteed that it is the same object. You believe it to be the same object, because you don't do anything special with reflection. By the same line of thought, I believe it to be the same object, because it's private, and assigned only once in constructor.

Discussions are great, because they open new perspectives. People can learn form each other, even if it's just a point of view. That's how we learn. 

Sometimes people cannot learn from each other, either by trying to enforce their own answer instead of exploring and learning, or for other reasons. This is when discussions tend to run in circles, with the same arguments over and over again. Such discussions bring no benefits to anyone.

I'm afraid we are running in circles.

**Comment by marcinmajkowski** [comment]
Without `final`, problem described above is present after constructor finishes. This is what I am trying to explain since the beginning of the discussion. My each and every post in this discussion is an another  approach to do that. Please just return to one of my earlier posts and look through the links I posted. Or read just a couple of the very first chapters in Java Concurrency in Practice and it should become clear.

You said you believe it's the same since its private and assigned in the constructor. But how do you know if somebody did not design that class such that for example `map` is replaced with the new one instead of cleared? Actually, it is what I initially expected when I saw you didn't use `final` there.

Since `final` fields may become inlined its a really bad idea to modify them with reflection. And if such use of reflection was involved into my codebase I would be aware of that. In my opinion it is such a marginal case that it should not affect my overall approach to coding.

**Comment by jakubnabrdalik** [comment]
@marcinmajkowski I have a simple solution to make this discussion fruitful: can you provide an example of a problem you are trying to solve?

My belief is, that once you try, you'll see it is not necessary to add final here. But I am very happy to be proven wrong (learning is always good). 

> But how do you know if somebody did not design that class such that for example map is replaced with the new one instead of cleared?

The same way you know it's not modified by reflection: 

> In my opinion it is such a marginal case that it should not affect my overall approach to coding.

**Comment by marcinmajkowski** [comment]
I presented you perfectly valid case where this reference could be modified without any smell (clearing repository) and you suggest that it is as marginal as using reflection to modify final fields.

I can't provide example since with current implementation of spring this problem doesn't exist but you're depending on an undocumented feature of framework. Concurrency issues often occur under heavy load and sometimes due to combination of various jvm settings, its considered impossible to test for them. I'm saying your code has bug due to Spring documentation and Java Language Specification.

Why then you don't need an example for race condition I fixed? It's the very same situation, except that you already understood race codition, and seems not to understand visibility and safe-publication issues.

You're welcome to commit deletion for `final` and merge this PR. I don't want bug introducing commit in my record.

**Comment by jakubnabrdalik** [comment]
> I presented you perfectly valid case where this reference could be modified without any smell (clearing repository) and you suggest that it is as marginal as using reflection to modify final fields.

Yeah, we do not have such a feature. In business software I've never had such a feature as clearing the whole repository.

> I can't provide example since with current implementation of spring this problem doesn't exist

Kind of my point

> but you're depending on an undocumented feature of framework.

This is where I think if you did try to implement this example, you'd see this makes no sense nonetheless

> I'm saying your code has bug due to Spring documentation and Java Language Specification.

Which you said cannot happen. I'd say it's not a bug if it cannot happen.

> Why then you don't need an example for race condition I fixed? 

Because it can happen

> It's the very same situation, except that you already understood race codition, and seems not to understand visibility and safe-publication issues.

Issues that as you say, cannot happen.

You seem to not understand, that adding "final" to an example in a situation that it does not protect you against anything (because "cannot happen", as you said) makes people follow a cargo cult of adding final everywhere, without understanding the reason. The same situation created the problem of "public" keyword being more popular than package-scope or private. I've also seen people adding final to every field (mutable or not) as a cargo cult. 

This is a trade-of: 
- protect code against something that cannot happen currently (and if did happen would show a bigger problem with design of the whole framework), but could in hypothetical situation, and risk a lot of people following the cargo cult
- do not protect code  against something that cannot happen, and risk people will NOT add final to a field that in their specific case (not a spring bean, not a repository), because they do not understand why

I choose the second solution. I think it's better not to create cargo cults, and make people think about every line of code, in the long run.

> You're welcome to commit deletion for final and merge this PR. I don't want bug introducing commit in my record.

That's very arrogant.  But, oh well.

**Comment by marcinmajkowski** [comment]
I'm sorry if it seemed arrogant to you but it really wasn't my intent. Spring implementation is free to change for undocumented behaviour without notice thats why you should not depend on it. Especially when bug fix is so simple. You can even comment on this `final` that it's due to safe-publication concerns and not due to expressing class design intent. I don't want to commit code that has bug in it.

When you decide to change scope of this bean, framework implementation may not longer protect you against this issue.

**Comment by jakubnabrdalik** [comment]
It's ok, I don't take it personal. This discussion was interesting before we started running in circles, and I'm grateful for it, and all the links you've provided. Thank you.

I'm disappointed we've failed to find a consensus. But that's life.

Did you know, by the way, that Spring implements  `@Named` against the specification? This should be prototype in any container following the specification, but Spring still does it differently (singleton by default). And I think their decision was correct.
So as you see, Spring is free to change any behaviour :)

**Comment by marcinmajkowski** [comment]
How I adhere to documented behavior probably comes from my c++ background where such issues are much more common. Environment changes are much more frequent there. There are even multiple standard library implementations. In Spring application development it is not such a big deal since there is (probably) one Spring implementation and most of us do not depend on any exotic JVM implementations so if something works in your environment then it is most likely completely all right.

We started running in circles because I thought that you suggest that I'm wrong and there is no safe-publication issue. But now (I hope so) the consensus is that indeed, class has safe-publication issue but since it does not affect you in context of this particular application, you're completely ok with it. Then I suggested that you should use `final` even if there was no concurrency-related issue, and probably I shouldn't do that since its up to you what coding practices you employ in your own project. And this was not the place for such discussion. Sorry for that.

The `@Named` case is indeed interesting. Luckily, surprising behavior related to this would become visible much earlier than any of concurrency-related issue.

Please let me know whether you prefer to fix race condition issue on your own or am I still welcome to contribute by separate PR fixing only that.

**Comment by sprzedwojski** [comment]
I ended up here by accident in 2020. Read through the whole discussion. It was really entertaining and educational, thank you guys 😄
